Fixing internal error in param-{find,show}

felipevolpone opened the pull request:
Fixing internal error in param-{find,show} with nonexistent object and showing properly error message.

Since PR #1013 probably won't be updated anymore, I created this one with previous work plus some additional changes.

https://pagure.io/freeipa/issue/7134

frasertweedale commented:
NACK.  The `param-{find,show}` commands work for me now, but `output-{find,show}` still
error in some cases:

```
[f25-2:~/dev/freeipa] [ pull/1033● ] ftweedal% ipa output-find fake         
ipa: ERROR: an internal error has occurred
[f25-2:~/dev/freeipa] [ pull/1033● ] ftweedal% ipa output-show fake fake
ipa: ERROR: an internal error has occurred
[f25-2:~/dev/freeipa] [ pull/1033● ] ftweedal% ipa output-show user-mod fake
ipa: ERROR: fake: output not found
[f25-2:~/dev/freeipa] [ pull/1033● ] ftweedal% ipa output-show user-mod value
  Name: value
  Documentation: The primary_key value of the entry, e.g. 'jdoe' for a user
  Type: str
```

frasertweedale commented:
@felipevolpone the latest commit has fixed the failure cases from my last comment but I still found a couple cases that caused internal error:

```
[f25-2:~/dev/freeipa] [ pull/1033● ] ftweedal% ipa output-find fake fake   
ipa: ERROR: an internal error has occurred
[f25-2:~/dev/freeipa] [ pull/1033● ] ftweedal% ipa param-find user-mod last
ipa: ERROR: an internal error has occurred
```

stlaz commented:
Travis is having issues with their new environment for builds. I pushed a temporary fix to `master`, feel free to rebase for the tests to pass properly.

rcritten commented:
A test should be added in ipatests/test_xmlrpc as well to ensure that this continues to function as expected, including both positive (e.g. user_show) and negative (e.g. fake) tests.

felipevolpone commented:
Tests added and patch updated to cover the behavior Fraser pointed out.

rcritten commented on ipatests/test_xmlrpc/test_schema_plugin.py:
```diff
@@ -0,0 +1,143 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
```
We use a standard contributors header now:

Copyright (C) 2017  FreeIPA Contributors see COPYING for license

rcritten commented on ipatests/test_xmlrpc/test_schema_plugin.py:
```diff
@@ -0,0 +1,143 @@
+        """Test param-find command"""
+
+        # right command without criteria
+        criteria = u'postalcode'
```
criteria is not used here

rcritten commented on ipatests/test_xmlrpc/test_schema_plugin.py:
```diff
@@ -0,0 +1,143 @@
+            assert (criteria in item['name'].lower() or
+                    criteria in item['doc'].lower()), item
+
+        # wrong command, right criteria
```
I guess you were on a roll but is there right criteria for a wrong command ;-)

rcritten commented on ipatests/test_xmlrpc/test_schema_plugin.py:
```diff
@@ -0,0 +1,143 @@
+        result = self.tracker.run_command('param_find', u'user-add', u'fake')
+        assert len(result['result']) == 0, result
+
+        # wrong command, wrong criteria
```
Same here. Maybe this is legitimate but I'd think that once the command is determined to be bad it doesn't matter what else was passed, right? I mean, how can one determine the "right" criteria for a command without having the command defined? In this case, postalcode makes sense for users but not for groups, so isn't correct in all cases anyway.

rcritten commented on ipatests/test_xmlrpc/test_schema_plugin.py:
```diff
@@ -0,0 +1,143 @@
+
+        # wrong command, wrong criteria
+        with pytest.raises(errors.NotFound):
+            result = self.tracker.run_command('param_show', u'fake', u'fake')
```
Two more similar criteria-related. I think a single test is sufficient.

rcritten commented on ipatests/test_xmlrpc/test_schema_plugin.py:
```diff
@@ -0,0 +1,143 @@
+class TestOutputFindAndShowCommand(XMLRPC_test):
+    """Test functionality of the ipa output-{find,show} command"""
+    tracker = Tracker()
+
```
Here is a philosophical question: is it worth comparing the output to something known or is it safe to assume that any output is ok? I'm not sure how much introspective is required and unless we want to test every single possible command's output it will still be hit or miss.

felipevolpone commented:
@rcritten I updated the PR with changes you proposed. Also, I added a simple check, to each command, in order to ensure it returns at least one right thing.

felipevolpone commented on ipatests/test_xmlrpc/test_schema_plugin.py:
```diff
@@ -0,0 +1,121 @@
+        result = self.tracker.run_command('param_find', u'user-add')
+        assert len(result['result']) != 0, result
+        expected = {'cli_name': 'login', 'name': 'uid', 'label': 'User login'}
+        assert set(expected.items()).issubset(set(result['result'][0].items()))
```
@rcritten if you think this line is too complicated; I can change it with separate comparisons, like:
```python
assert result['result'][0]['cli_name'] == 'login'
assert result['result'][0]['name'] == 'uid'
```

rcritten replied:
Ideally it would be broken out so in case of failure it would be clear which value failed the assertion.

rcritten commented:
Works great.

Tiboris commented:
master:

* 7ab49dda6db52c0960f50f1744aa19dc70d3c62f schema: Fix internal error in param-{find,show} with nonexistent object
* 3822120077d5bd0f951560d3181654d1f28fca4e Fixing param-{find,show} and output-{find,show} commands
